#!/usr/bin/env bash


# First requires install of autometa...
# Can do by cloning and installing manually or through conda.
# If manually installing Autometa, you will also need to install env.

email="erees@wisc.edu"

template="$HOME/sponge_paper/sponge_paper/src/features/template.mf"
features="$HOME/sponge_paper/sponge_paper/src/features"
metagenomes=`ls $HOME/sponge_paper/sponge_paper/data/raw/assemblies/*.fasta`
interim="$HOME/sponge_paper/sponge_paper/data/interim"
filtered_assemblies="${interim}/assemblies"
binnings="${interim}/binning"
kmer_norm="ilr"
kmer_embed="bhsne"

batch="${features}/run_makeflows.sh"
# overwrite if it already exists...
if [ -f $batch ];then
    echo "Removing existing ${batch}"
    rm $batch
fi

cat > $batch <<- EOM
# Sponge Makeflows generated by
# $HOME/sponge_paper/sponge_paper/src/make_sponge_makeflows.sh

EOM


logs="${binnings}/logs"

if [ ! -d $logs ];then
    mkdir -p $logs
fi

# example: ${mg} == "FL2015_4.fasta"
for mg in $metagenomes;do
    metagenome=$(basename ${mg/.fasta/})
    makeflow="${features}/${metagenome}.mf"
    # Copy template.mf to FL2015_4.mf
    cp ${template} ${makeflow}
    # edit FL2015_4.mf inputs and outputs filepaths
    sed -i "s,metagenome.fasta,${mg},g" ${makeflow}
    # FILTERED
    filtered_assembly="${filtered_assemblies}/${metagenome}.filtered.fna"
    sed -i "s,metagenome.filtered.fna,${filtered_assembly},g" ${makeflow}
    # KMERS
    kmers="${binnings}/${metagenome}.kmers."
    sed -i "s,kmers\.,${kmers},g" ${makeflow}
    # COVERAGES
    coverages="${binnings}/${metagenome}.coverages."
    sed -i "s,coverages\.,${coverages},g" ${makeflow}
    # ORFS
    orfs="${binnings}/${metagenome}.orfs."
    sed -i "s,orfs\.,${orfs},g" ${makeflow}
    # TAXONOMY
    taxonomy="${binnings}/${metagenome}.taxonomy."
    sed -i "s,taxonomy\.,${taxonomy},g" ${makeflow}
    # BLASTP
    blastp="${binnings}/${metagenome}.blastp."
    sed -i "s,blastp\.,${blastp},g" ${makeflow}
    # HITS
    hits="${binnings}/${metagenome}.hits."
    sed -i "s,hits\.,${hits},g" ${makeflow}
    # LCA
    lca="${binnings}/${metagenome}.lca."
    sed -i "s,lca\.,${lca},g" ${makeflow}
    # BACTERIA HMMSCAN
    hmmscan="${binnings}/${metagenome}.bacteria.hmmscan."
    sed -i "s,bacteria\.hmmscan\.,${hmmscan},g" ${makeflow}
    # ARCHAEA HMMSCAN
    hmmscan="${binnings}/${metagenome}.archaea.hmmscan."
    sed -i "s,archaea\.hmmscan\.,${hmmscan},g" ${makeflow}
    # BACTERIA MARKERS
    markers="${binnings}/${metagenome}.bacteria.markers."
    sed -i "s,bacteria\.markers\.,${markers},g" ${makeflow}
    # ARCHAEA MARKERS
    markers="${binnings}/${metagenome}.archaea.markers."
    sed -i "s,archaea\.markers\.,${markers},g" ${makeflow}
    # BACTERIA BINNING
    binning="${binnings}/${metagenome}.bacteria.binning."
    sed -i "s,bacteria\.binning\.,${binning},g" ${makeflow}
    # ARCHAEA BINNING
    binning="${binnings}/${metagenome}.archaea.binning."
    sed -i "s,archaea\.binning\.,${binning},g" ${makeflow}
    # BACTERIA UNCLUSTERED RECRUITMENT
    recruitment="${binnings}/${metagenome}.bacteria.recruitment."
    sed -i "s,bacteria.unclustered_recruitment\.,${recruitment},g" ${makeflow}
    # ARCHAEA UNCLUSTERED RECRUITMENT
    recruitment="${binnings}/${metagenome}.archaea.recruitment."
    sed -i "s,archaea.unclustered_recruitment\.,${recruitment},g" ${makeflow}

    #### Now we start writing our slurm script ###
    sponge_batch="${features}/run_${metagenome}_makeflow.sh"
    # overwrite if it already exists...
    if [ -f $sponge_batch ];then
        echo "Removing existing ${sponge_batch}"
        rm $sponge_batch
    fi

    cat >> $sponge_batch <<- EOM
#!/bin/bash
#SBATCH --partition=queue
#SBATCH --time=14-00:00:00
#SBATCH -N 1 # Nodes
#SBATCH -n 1 # Tasks
#SBATCH --cpus-per-task=1
EOM
    # Notice we are redirecting slurm stderr and stdout to a logs directory...
    #  You will need to make this directory if it does not exists where you are submitting
    slurm_stderr="#SBATCH --error=logs/%J.run_${metagenome}_makeflow.err"
    slurm_stdout="#SBATCH --output=logs/%J.run_${metagenome}_makeflow.out"
    echo $slurm_stderr >> $sponge_batch
    echo $slurm_stdout >> $sponge_batch
    cat >> $sponge_batch <<- EOM
# Sponge Makeflows generated by
# $HOME/sponge_paper/sponge_paper/src/make_sponge_makeflows.sh

EOM

    ### Setup our makeflow monitoring directory (This should be unique from other directories)
    metagenome_logs="${logs}/${metagenome}"
    if [ ! -d $metagenome_logs ];then
        mkdir -p $metagenome_logs
    fi
    summarylog="${logs}/${metagenome}.summary.log"

    # command="makeflow -N 1 -n 1 -c \${CORES} --mem=\${MEMORY}M --email=${email} --summary-log=${summarylog} --monitor=${metagenome_logs} -T slurm ${makeflow}"
    command="makeflow --local-cores=40 --email=${email} --summary-log=${summarylog} --monitor=${metagenome_logs} -T slurm ${makeflow}"
    echo $command >> ${sponge_batch}

    ### Now add sbatch metagenome.sh to run_makeflows.sh for submitting all to queue at once.
    # sbatch <path/to/metagenome.sh>
    echo "sbatch ${sponge_batch}" >> ${batch}

done